api/webhook.go

```diff
@@ -0,0 +1,202 @@
+	}
+	w.Header().Set("Content-Type", "application/json")
+	return json.NewEncoder(w).Encode(webhook)
+}
+
+// title: webhook create
+// path: /events/webhooks
+// method: POST
+// responses:
+//   200: Webhook created
+//   401: Unauthorized
+//   400: Invalid webhook
+//   409: Webhook already exists
+func webhookCreate(w http.ResponseWriter, r *http.Request, t auth.Token) error {
+	r.ParseForm()
+	dec := form.NewDecoder(nil)
+	dec.IgnoreUnknownKeys(true)
+	dec.IgnoreCase(true)
+	var webhook eventTypes.WebHook
+	err := dec.DecodeValues(&webhook, r.Form)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return err
+	}
+	if webhook.TeamOwner == "" {
+		webhook.TeamOwner, err = autoTeamOwner(t, permission.PermWebhookCreate)
+		if err != nil {
+			return err
+		}
+	}
+	ctx := permission.Context(permission.CtxTeam, webhook.TeamOwner)
+	if !permission.Check(t, permission.PermWebhookCreate, ctx) {
+		return permission.ErrUnauthorized
+	}
+	evt, err := event.New(&event.Opts{
+		Target:     event.Target{Type: event.TargetTypeWebHook, Value: webhook.Name},
+		Kind:       permission.PermWebhookCreate,
+		Owner:      t,
+		CustomData: event.FormToCustomData(r.Form),
+		Allowed:    event.Allowed(permission.PermWebhookReadEvents, ctx),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		evt.Done(err)
+	}()
+	err = servicemanager.WebHook.Create(webhook)
+	if err == eventTypes.ErrWebHookAlreadyExists {
+		w.WriteHeader(http.StatusConflict)
+	} else if _, ok := err.(*errors.ValidationError); ok {
+		w.WriteHeader(http.StatusBadRequest)
+	}
+	return err
+}
+
+// title: webhook update
+// path: /events/webhooks/{name}
+// method: PUT
+// responses:
+//   200: Webhook updated
+//   401: Unauthorized
+//   400: Invalid webhook
+//   404: Webhook not found
+func webhookUpdate(w http.ResponseWriter, r *http.Request, t auth.Token) error {
+	r.ParseForm()
+	dec := form.NewDecoder(nil)
+	dec.IgnoreUnknownKeys(true)
+	dec.IgnoreCase(true)
+	var webhook eventTypes.WebHook
+	err := dec.DecodeValues(&webhook, r.Form)
+	if err != nil {
+		return &errors.HTTP{Code: http.StatusBadRequest, Message: fmt.Sprintf("unable to parse webhook: %v", err)}
+	}
+	webhook.Name = r.URL.Query().Get(":name")
+	ctx := permission.Context(permission.CtxTeam, webhook.TeamOwner)
+	if !permission.Check(t, permission.PermWebhookUpdate, ctx) {
+		return permission.ErrUnauthorized
+	}
+	evt, err := event.New(&event.Opts{
+		Target:     event.Target{Type: event.TargetTypeWebHook, Value: webhook.Name},
+		Kind:       permission.PermWebhookUpdate,
+		Owner:      t,
+		CustomData: event.FormToCustomData(r.Form),
+		Allowed:    event.Allowed(permission.PermWebhookReadEvents, ctx),
+	})
+	if err != nil {
+		return err
+	}
+	defer func() {
+		evt.Done(err)
+	}()
+	err = servicemanager.WebHook.Update(webhook)
+	if err == eventTypes.ErrWebHookNotFound {
+		w.WriteHeader(http.StatusNotFound)
+	} else if _, ok := err.(*errors.ValidationError); ok {
+		w.WriteHeader(http.StatusBadRequest)
```

Comment:
This is already handled by the middleware: https://github.com/tsuru/tsuru/blob/master/api/middleware.go#L119

Reply:
👍